[1.11.2] Fixed Zombie Villagers not setting and getting professions correctly

Madsthunder opened the pull request:
This pr does three things:

1. It changes `EntityZombie#onKillEntity (line 409)` to 
`EntityZombieVillager.setForgeProfession(entityvillager.getProfessionForge)`
**instead of**
`EntityZombieVillager.setProfession(entityvillager.getProfession)`,
since `EntityVillager#getProfession` can only return 6 values, rather than 1024.

2. It fixes `EntityZombieVillager` to handle `VillagerProfession`s correctly by allowing it to read, write, and update `EntityZombieVillager.prof` through `EntityZombieVillager#readEntityFromNBT`, `#writeEntityToNBT`, and `#notifyDataManagerChange`.

3. It changes `RenderZombieVillager` to use `EntityZombieVillager.getForgeProfession#getZombieSkin` instead of a `switch` block, similar to `RenderVillager`.

Madsthunder committed:
Fix zombie villagers not setting professions correctly

Madsthunder committed:
Fix the conversion from Zombie Villager to Villager

mezz commented on patches/minecraft/net/minecraft/client/renderer/entity/RenderZombieVillager.java.patch:
```diff
@@ -0,0 +1,26 @@
+-            default:
+-                return field_110864_q;
+-        }
++        return p_110775_1_.getForgeProfession().getZombieSkin();
```
This patch can be made one line, put 
`if (true) return p_110775_1_.getForgeProfession().getZombieSkin();` 
before the switch statement, and do not delete anything.

mezz reviewed (changes requested):
Looks good, only one issue with minimizing patches.
I'll test out this when I have some more time, the test mod looks like it'll be nice to have.

Madsthunder committed:
Apply some suggestions (1)

williewillus commented on patches/minecraft/net/minecraft/client/renderer/entity/RenderZombieVillager.java.patch:
```diff
@@ -0,0 +1,10 @@
+ 
+     protected ResourceLocation func_110775_a(EntityZombieVillager p_110775_1_)
+     {
++    	if (true) return p_110775_1_.getForgeProfession().getZombieSkin();
```
mixed tab + space

Madsthunder replied:
Argh, I need to fix my workspace

Madsthunder committed:
Apply some suggestions (2)

Madsthunder commented:
Hm, Travis is fixed

bs2609 commented:
This needs updating to resolve conflicts

mezz commented:
@Madsthunder please update this to resolve conflicts

bs2609 commented on patches/minecraft/net/minecraft/entity/monster/EntityZombie.java.patch:
```diff
@@ -56,6 +56,17 @@
                              entityzombie.func_110148_a(field_110186_bp).func_111121_a(new AttributeModifier("Zombie reinforcement callee charge", -0.05000000074505806D, 0));
+
+
+@@ -394,7 +406,7 @@
```
These blank lines don't look right...

bs2609 commented on patches/minecraft/net/minecraft/entity/monster/EntityZombie.java.patch:
```diff
@@ -64,4 +75,4 @@
          }
  
-         if (p_180482_2_ instanceof EntityZombie.GroupData)
+         if (p_180482_2_ instanceof EntityZombie.GroupData)
```
Also this line shouldn't have changed

Madsthunder replied:
Hmm, strange... I don't see how the line changed

Madsthunder committed:
Fixed some merging errors.

mezz commented:
Tested and if I save and reload, the modded zombies revert to vanilla zombies.

Madsthunder commented:
Huh, I guess I forgot to patch `readEntityFromNBT`. I'll get around to fixing that sometime tomorrow.

bs2609 commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -20,13 +28,36 @@
 +        return this.prof;
 +    }
 +
++    public void func_184206_a(DataParameter<?> key)
```
In `EntityVillager`, this is annotated `@Override`, do you think that should be done here too?

Madsthunder replied:
Might as well do so

Madsthunder committed:
Fixed Zombie Villager not reading forge profession

mezz commented:
Tested and looks good.
I converted some vanilla and the modded villagers into zombies and back, saving and reloading at each stage to make sure they kept the occupation.

LexManos commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -8,7 +8,28 @@
      public int func_190736_dl()
-@@ -85,6 +86,21 @@
+@@ -64,6 +65,7 @@
+     {
```
None of the changes in this file should be needed, they were only hacks because the old system mixed in profession and normal villager types.
Now that that is no longer the case, we don't have to special case the profession string.
What MAY need to happen is us going in and removing any explicit limitations to the number 6, the number of vanilla professions.

Madsthunder replied:
Should I go ahead and do it in this pr, so it won't be as much of an annoyance in the future?

LexManos commented on patches/minecraft/net/minecraft/entity/monster/EntityZombie.java.patch:
```diff
@@ -56,6 +56,15 @@
+             this.field_70170_p.func_72900_e(entityvillager);
+             entityzombievillager.func_180482_a(this.field_70170_p.func_175649_E(new BlockPos(entityzombievillager)), new EntityZombie.GroupData(false));
+-            entityzombievillager.func_190733_a(entityvillager.func_70946_n());
++            entityzombievillager.setForgeProfession(entityvillager.getProfessionForge());
```
again, shouldn't be needed as setProfession(int) is in the full profession id space and doesn't need translations.

Madsthunder replied:
Should I keep it for this pr? Or should I remove all the cycles to 6 in `EntityVillager` and `EntityZombieVillager`?

LexManos replied:
Yes, as everything except getProfession should support ids > 6

Madsthunder replied:
Actually `getProfession` in these two classes are the only places where ids > 6 aren't supported. And since `getProfession` should still return ids <= 6 then the change above is still needed.

LexManos replied:
What I meant is that you should fix getProfession %6 there is no reason for it to be doing that anymore.
Everywhere else supports professions > 6 no reason the getter should be screwing things up.

LexManos commented on patches/minecraft/net/minecraft/client/renderer/entity/RenderZombieVillager.java.patch:
```diff
@@ -0,0 +1,10 @@
+     protected ResourceLocation func_110775_a(EntityZombieVillager p_110775_1_)
+     {
++        if (true) return p_110775_1_.getForgeProfession().getZombieSkin();
+         switch (p_110775_1_.func_190736_dl())
```
This is actually legit, seems i skipped/forgot about this when i was doing it... could of sworn i did skins tho...

LexManos commented on src/test/java/net/minecraftforge/test/VillagerProfessionTest.java:
```diff
@@ -0,0 +1,21 @@
+    public static void registerVillagers(RegistryEvent.Register<VillagerProfession> event)
+    {
+        VillagerProfession profession = new VillagerProfession("professiontest:test_villager", "professiontest:textures/entity/test_villager.png", "professiontest:textures/entity/zombie_test_villager.png");
+        new VillagerCareer(profession, "professiontest:test_villager");
```
where are those textures from? Dont like sticking random images in our repo without knowing the source/copyrights.

Madsthunder replied:
The textures are just random textures I made in paint.net, nothing special.

liach commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -8,7 +8,28 @@
++        if (p_70037_1_.func_74764_b("ProfessionName"))
++        {
++            net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession p = net.minecraftforge.fml.common.registry.VillagerRegistry.instance().getRegistry().getValue(new net.minecraft.util.ResourceLocation(p_70037_1_.func_74779_i("ProfessionName")));
++            if (p == null) p = net.minecraftforge.fml.common.registry.VillagerRegistry.instance().getRegistry().getValue(new net.minecraft.util.ResourceLocation("minecraft:farmer"));
```
I suggest to cache this somewhere

Madsthunder replied:
You're referring to the farmer, right?

liach replied:
Yes. Cache the profession somewhere since it is used in many places.

Madsthunder replied:
How about an `@ObjectHolder` in `VillagerRegistry`?

LexManos replied:
Actually would be a good idea to have a constants class like everything else that uses a registry. Should talk to Mojang about that...

liach commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -20,13 +41,37 @@
 +    }
 +
++    @Override
++    public void func_184206_a(DataParameter<?> key)
```
I suggest to move this block to the end and add 
```java
/*---------------Forge Start--------------------*/
```

Madsthunder replied:
Should I do this in EntityVillager too? Or should that be for another PR?

liach replied:
Just for this method now

liach replied:
Done. Great!

liach commented on src/test/java/net/minecraftforge/test/VillagerProfessionTest.java:
```diff
@@ -0,0 +1,21 @@
+import net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerCareer;
+import net.minecraftforge.fml.common.registry.VillagerRegistry.VillagerProfession;
+
+@Mod(modid = "professiontest", name = "ProfessionTest2000", version = "1.0", acceptableRemoteVersions = "*")
```
This mod should be required by both sides since it adds new textures

LexManos replied:
no it shouldnt. The vanilla client *should* be able to connect to servers with this mod it would just have zombies/villagers with the wrong professions.

liach replied:
Well, didn't know how exactly this worked

liach commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -20,13 +41,37 @@
+     {
+         EntityVillager entityvillager = new EntityVillager(this.field_70170_p);
+         entityvillager.func_82149_j(this);
+-        entityvillager.func_70938_b(this.func_190736_dl());
```
How about deprecating the old method?

Madsthunder committed:
Applied some suggestions (3)

Added a FARMER constant in VillagerRegistry
Moved some extra methods in EntityZombieVillager to a forge start-end
block
Deprecated EntityZombieVillager#getProfession(int)

liach commented on patches/minecraft/net/minecraft/entity/monster/EntityZombieVillager.java.patch:
```diff
@@ -20,13 +57,26 @@
++        {
++            this.prof = net.minecraftforge.fml.common.registry.VillagerRegistry.getById(this.func_190736_dl());
++            if (this.prof == null)
++                return net.minecraftforge.fml.common.registry.VillagerRegistry.getById(0); //Farmer
```
```java
return net.minecraftforge.fml.common.registry.VillagerRegistry.FARMER;
```

Madsthunder committed:
Apply Some Suggestions(3.1)

Added a FARMER constant in VillagerRegistry

liach commented on jsons/1.11.2.json:
```diff
@@ -558,6 +558,6 @@
     "minecraftArguments": "--username ${auth_player_name} --version ${version_name} --gameDir ${game_directory} --assetsDir ${assets_root} --assetIndex ${assets_index_name} --uuid ${auth_uuid} --accessToken ${auth_access_token} --userType ${user_type} --versionType ${version_type}",
     "minimumLauncherVersion": 18,
     "releaseTime": "2016-12-21T09:29:12+00:00",
-    "time": "2017-02-07T13:18:39+00:00",
```
Remove this file

liach replied:
So, you should copy paste the original version onto the changed version.

Madsthunder committed:
Applied Some Suggestions (3.2)

Removed 1.11.2.json

Madsthunder commented:
@LexManos I accidentally deleted 1.11.2.json, will it affect 1.11.2.json on 1.11.x? If so, should I restore it to the newest version or the current version?

LexManos commented:
Just revert it in git... you should never make Prs with things that you don't change...

Madsthunder committed:
Revert "Applied Some Suggestions (3.2)"

This reverts commit 20df0ca51e661e1c0b396bfe3801ab7e459b869b.

Madsthunder committed:
Applied Some Suggestions (3.3)

Updated 1.11.2.json to the branch default

Madsthunder committed:
Applied Some Suggestions (4)

Removed the limit of 6 in the #getProfession methods of EntityVillager
and EntityZombieVillager
Changed the reference to EntityZombieVillager#setForgeProfession and
EntityVillager#getProfessionForge in EntityZombie#killEntity to
#setProfession and #getProfession

mezz commented on patches/minecraft/net/minecraft/entity/monster/EntityZombie.java.patch:
```diff
@@ -56,12 +56,3 @@
-         if (p_180482_2_ == null)
-         {
--            p_180482_2_ = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < 0.05F);
-+            p_180482_2_ = new EntityZombie.GroupData(this.field_70170_p.field_73012_v.nextFloat() < net.minecraftforge.common.ForgeModContainer.zombieBabyChance);
```
It looks like you removed this existing patch accidentally.

Madsthunder replied:
Oops

Madsthunder committed:
Applied Some Suggestions (4.1)

Restored missing patch from #3691

mezz commented:
Going to test this again. It's not clear what has changed because of the git derps.

mezz commented:
Gave it a quick test and everything appears to still be working correctly.

Madsthunder commented:
Should I move `VillagerRegistry.FARMER` to a `VillagerRegistry.Professions` class, similar to #3862?

mezz commented:
Yes, that looks like a cleaner way to do it.

liach commented:
Hail!
